woocommerce/woocommerce
# Improve Accessibility in Product Collection Navigation

### Changes proposed in this Pull Request:





This PR introduces improvements to enhance accessibility and user experience. Specifically, it adds new context properties to manage accessibility messages during the navigation process, including messages for loading and when a page has loaded. These changes ensure that screen reader users receive appropriate feedback during navigation.

Changes made:
- Added `accessibilityMessage`, `accessibilityLoadingMessage`, and `accessibilityLoadedMessage` properties to the `ProductCollectionStoreContext`. These properties store messages to be announced by screen readers during different stages of page navigation.
- Implemented logic in the product collection store to update the `accessibilityMessage` during the start of navigation (showing a loading message) and upon completion (showing a loaded message).
- Utilized a technique to ensure that consecutive identical messages are still announced by screen readers, by appending a no-break space to the message if it is the same as the previous one. This follows a pattern similar to the `@wordpress/a11y` package.
- In `ProductCollection.php`, enhanced the block's HTML output to include these new accessibility messages and integrated them with the existing interactive data attributes. This ensures that the front-end components are fully prepared to handle these accessibility enhancements.
- Added a new `div` with `class="screen-reader-text"` and `aria-live="polite"` attributes, which dynamically displays the accessibility message based on the navigation state. This div complements the visual pagination animation with an accessibility-focused mechanism for announcing the page load states to screen reader users.

By addressing accessibility considerations with these enhancements, we're making Product Collection navigation more inclusive and user-friendly for all users.

Closes #44394 



### How to test the changes in this Pull Request:



1. Add a Product Collection block to a new post & choose any collection.
2. Adjust the items per page in the toolbar controls to "3" to ensure only three products are displayed per page, creating multiple pages for navigation.
3. Publish the post & go to Frontend.
4. Enable VoiceOver (here are [instructions for the MacOS](https://support.apple.com/guide/voiceover/turn-voiceover-on-or-off-vo2682/mac#:~:text=Choose%20Apple%20menu%20%3E%20System%20Settings,turn%20VoiceOver%20on%20or%20off.))
    1. Tip: You can use shortcut “Cmd + F5” to enable VoiceOver
5. Verify that changing the page should announce:
    1. “Loading page, please wait.” (When the page is loading)
    2. “Page Loaded.” (When the page is loaded)

**Tip:** Consider throttling the network speed to "Slow 3G" to simulate slower loading times.



### Changelog entry




<details>

#### Significance




#### Type




#### Message 
Improve Accessibility in Product Collection Navigation

#### Comment 

</details>
